File: go/kan/rift.go
package capitan

type Rift struct {
    runRift func(bac func(b interface{}) (interface{}, interface{}), gc interface{}) interface{}
}

func (r Rift) ContraMap(f func(a interface{}) interface{}) Rift {
    return Rift{
        runRift: func(bac func(b interface{}) (interface{}, interface{}), gc interface{}) interface{} {
            return r.runRift(func(b interface{}) (interface{}, interface{}) {
                a, c := bac(b)
                return f(a), c
            }, gc)
        },
    }
}

func Divide(f func(a interface{}) interface{}, g Rift, h Rift) Rift {
    return Rift{
        runRift: func(bac func(b interface{}) (interface{}, interface{}), gc interface{}) interface{} {
            return g.runRift(func(a interface{}) (interface{}, interface{}) {
                c, d := bac(a)
                return f(c), d
            }, gc)
        },
    }
}

File: rust/src/kan/rift.rs
struct Rift<G, H, A>(Box<dyn Fn(Box<dyn Fn(B) -> (A, C)>, G) -> H>);

impl<G, H, A, B, C> Rift<G, H, A>
where
    G: Clone,
    H: Clone,
{
    fn contramap<F>(self, f: F) -> Rift<G, H, B>
    where
        F: Fn(B) -> A + 'static,
    {
        Rift(Box::new(move |bac: Box<dyn Fn(B1) -> (B, C)>, g: G| {
            (self.0)(Box::new(move |b: B1| {
                let (a, c) = bac(b);
                (f(a), c)
            }), g.clone())
        }))
    }
}

impl<G, A, B, C> Rift<G, G, A>
where
    G: Clone,
{
    fn divide<F>(f: F, g: Rift<G, G, A>, h: Rift<G, G, B>) -> Rift<G, G, C>
    where
        F: Fn(A) -> B + 'static,
    {
        Rift(Box::new(move |bac: Box<dyn Fn(B1) -> (C, D)>, g: G| {
            g.0(Box::new(move |a: B1| {
                let (c, d) = bac(a);
                (f(c), d)
            }), g.clone())
        }))
    }
}